contracts/axelarnet-gateway/src/contract/execute.rs

```diff
@@ -1,76 +1,177 @@
-};
-use error_stack::{report, Result, ResultExt};
+use axelar_wasm_std::{FnExt, IntoContractError};
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{Addr, DepsMut, HexBinary, Response, Storage, Uint256};
+use error_stack::{bail, ensure, report, Result, ResultExt};
+use itertools::Itertools;
 use router_api::client::Router;
 use router_api::{Address, ChainName, CrossChainId, Message};
 use sha3::{Digest, Keccak256};
 
-use crate::contract::Error;
+use crate::clients::CrossChainExecutor;
 use crate::events::AxelarnetGatewayEvent;
-use crate::executable::AxelarExecutableClient;
-use crate::state::{self};
+use crate::state;
+use crate::state::Config;
+
+#[derive(thiserror::Error, Debug, IntoContractError)]
+pub enum Error {
+    #[error("failed to save executable message")]
+    SaveExecutableMessage,
+    #[error("failed to access executable message")]
+    ExecutableMessageAccess,
+    #[error("message with ID {0} does not match the expected message")]
+    MessageMismatch(CrossChainId),
+    #[error("failed to mark message with ID {0} as executed")]
+    MarkExecuted(CrossChainId),
+    #[error("expected destination chain {expected}, got {actual}")]
+    InvalidDestination {
+        expected: ChainName,
+        actual: ChainName,
+    },
+    #[error("failed to generate cross chain id")]
+    CrossChainIdGeneration,
+    #[error("unable to load the contract config")]
+    ConfigAccess,
+    #[error("unable to save the message before routing")]
+    SaveRoutableMessage,
+    #[error("invalid cross-chain id")]
+    InvalidCrossChainId,
+    #[error("failed to create executor")]
+    CreateExecutor,
+    #[error("not allowed to execute payload")]
+    PayloadNotApproved,
+    #[error("unable to generate event index")]
+    EventIndex,
+    #[error("invalid source address {0}")]
+    InvalidSourceAddress(Addr),
+}
+
+#[cw_serde]
+pub struct CallContractData {
+    pub destination_chain: ChainName,
+    pub destination_address: Address,
+    pub payload: HexBinary,
+}
+
+impl CallContractData {
+    pub fn into_message(self, id: CrossChainId, source_address: Address) -> Message {
+        Message {
+            cc_id: id,
+            source_address,
+            destination_chain: self.destination_chain,
+            destination_address: self.destination_address,
+            payload_hash: Keccak256::digest(self.payload.as_slice()).into(),
+        }
+    }
+}
 
-#[allow(clippy::too_many_arguments)]
 pub fn call_contract(
-    store: &mut dyn Storage,
+    storage: &mut dyn Storage,
     block_height: u64,
-    router: &Router,
-    chain_name: ChainName,
     sender: Addr,
-    destination_chain: ChainName,
-    destination_address: Address,
-    payload: HexBinary,
+    call_contract: CallContractData,
 ) -> Result<Response, Error> {
-    let counter = state::increment_msg_counter(store).change_context(Error::InvalidStoreAccess)?;
+    let payload = call_contract.payload.clone();
```

Comment:
feels unnecessary to have to clone the whole payload, just because of how `into_message` works even though it only uses it as a slice. Maybe it takes a reference instead, or change it to something like `into_message_and_payload` that returns `(msg, payload)`.